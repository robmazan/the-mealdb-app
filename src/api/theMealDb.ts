const API_KEY = process.env.REACT_APP_MEALDB_API_KEY || 1;

export interface Meal {
  idMeal: number;
  strMeal: string;
  strDrinkAlternate: string;
  strCategory: string;
  strArea: string;
  strInstructions: string;
  strMealThumb: string;
  strTags: string;
  strYoutube: string;
  strIngredient1: string;
  strIngredient2: string;
  strIngredient3: string;
  strIngredient4: string;
  strIngredient5: string;
  strIngredient6: string;
  strIngredient7: string;
  strIngredient8: string;
  strIngredient9: string;
  strIngredient10: string;
  strIngredient11: string;
  strIngredient12: string;
  strIngredient13: string;
  strIngredient14: string;
  strIngredient15: string;
  strIngredient16: string;
  strIngredient17: string;
  strIngredient18: string;
  strIngredient19: string;
  strIngredient20: string;
  strMeasure1: string;
  strMeasure2: string;
  strMeasure3: string;
  strMeasure4: string;
  strMeasure5: string;
  strMeasure6: string;
  strMeasure7: string;
  strMeasure8: string;
  strMeasure9: string;
  strMeasure10: string;
  strMeasure11: string;
  strMeasure12: string;
  strMeasure13: string;
  strMeasure14: string;
  strMeasure15: string;
  strMeasure16: string;
  strMeasure17: string;
  strMeasure18: string;
  strMeasure19: string;
  strMeasure20: string;
  strSource: string;
  dateModified: string;
}

export interface Category {
  idCategory: number;
  strCategory: string;
  strCategoryThumb: string;
  strCategoryDescription: string;
}

export const searchByName = (name: string): Promise<Meal[]> =>
  fetch(
    `https://www.themealdb.com/api/json/v1/${API_KEY}/search.php?s=${encodeURIComponent(
      name
    )}`
  )
    .then(res => res.json())
    .then(res => res.meals);

export const searchByFirstLetter = (firstLetter: string): Promise<Meal[]> =>
  fetch(
    `https://www.themealdb.com/api/json/v1/${API_KEY}/search.php?f=${encodeURIComponent(
      firstLetter
    )}`
  )
    .then(res => res.json())
    .then(res => res.meals);

export const findById = (id: number): Promise<Meal> =>
  fetch(
    `https://www.themealdb.com/api/json/v1/${API_KEY}/lookup.php?i=${encodeURIComponent(
      id
    )}`
  )
    .then(res => res.json())
    .then(res => res.meals[0]);

export const fetchRandom = (): Promise<Meal> =>
  fetch(`https://www.themealdb.com/api/json/v1/${API_KEY}/random.php`)
    .then(res => res.json())
    .then(res => res.meals[0]);

export const fetchCategories = (): Promise<Category[]> =>
  fetch(`https://www.themealdb.com/api/json/v1/${API_KEY}/categories.php`)
    .then(res => res.json())
    .then(res => res.categories);
