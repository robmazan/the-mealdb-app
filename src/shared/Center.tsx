import React from "react";
import styles from "./Center.module.css";

const Center: React.FC = ({ children }) => {
  return <div className={styles.wrapper}>{children}</div>;
};

export default Center;
