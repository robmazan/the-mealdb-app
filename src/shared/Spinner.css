@keyframes spinning {
  0% {
    stroke-dashoffset: 0;
    stroke-dasharray: 150 100;
  }
  50% {
    stroke-dasharray: 1 250;
  }
  100% {
    stroke-dashoffset: 500;
    stroke-dasharray: 150 100;
  }
}

.spinner-circle {
  animation: 2s linear spinning infinite;
}
