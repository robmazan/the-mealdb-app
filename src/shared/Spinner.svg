<svg width="200px" height="200px" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid">
  <circle class="spinner-circle" cx="50" cy="50" r="40" stroke="#ffe4c4" fill="none" stroke-width="6" stroke-linecap="round"></circle>
</svg>
