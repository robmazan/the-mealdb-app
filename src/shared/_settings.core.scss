$fonts: (
    pacifico: ('Pacifico', cursive)
);

@function get-font($font) {
    @return map-get($fonts, $font);
}

$colors: (
    firefly: #113333,
    negroni: #ffe4c4
);

@function get-color($color) {
    @return map-get($colors, $color);
}
