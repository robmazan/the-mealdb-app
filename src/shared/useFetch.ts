import { useState, useEffect } from "react";

export enum LoadingState {
  PENDING,
  ERROR,
  DONE
}

const useFetch = <T, M = T>(
  url: string,
  resultMapper?: (res: T) => M,
  initialData?: M
): [M | undefined, LoadingState, Error | null] => {
  const [data, setData] = useState<M | undefined>(initialData);
  const [loadingState, setLoadingState] = useState<LoadingState>(
    LoadingState.PENDING
  );
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    let mounted = true;
    const abortController = new AbortController();

    const doFetch = async () => {
      try {
        const res = await fetch(url, { signal: abortController.signal });
        if (!res.ok) {
          throw new Error(
            `Error during fetching "${url}! HTTP ${res.status}: ${res.statusText}"`
          );
        }
        let data = await res.json();

        if (resultMapper) {
          data = resultMapper(data);
        }

        if (mounted) {
          setData(data);
          setLoadingState(LoadingState.DONE);
        }
      } catch (e) {
        setError(e);
        setLoadingState(LoadingState.ERROR);
      }
    };

    doFetch();

    const cleanup = () => {
      mounted = false;
      abortController.abort();
    };

    return cleanup;
  }, [url, resultMapper]);

  return [data, loadingState, error];
};

export default useFetch;
