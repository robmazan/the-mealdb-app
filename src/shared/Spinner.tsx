import { ReactComponent as Spinner } from "./Spinner.svg";
import "./Spinner.css";

export default Spinner;
