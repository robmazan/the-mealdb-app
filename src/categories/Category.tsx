import React from "react";
import { useParams } from "react-router";
import styles from "./Category.module.scss";
import Thumbnail from "./Thumbnail";
import { LoadingState } from "../shared/useFetch";
import { useMealExcerptsOfCategory } from "../api/theMealDb";
import Center from "../shared/Center";
import Spinner from "../shared/Spinner";

const Category: React.FC = () => {
  const { categoryName } = useParams<{ categoryName: string }>();
  const [meals, loadingState, loadError] = useMealExcerptsOfCategory(
    categoryName
  );

  document.title = `The Meal DB: ${categoryName}`;

  switch (loadingState) {
    case LoadingState.DONE:
      return (
        <>
          <h1 className={styles.pageTitle}>{categoryName}</h1>
          <ul className={styles.list}>
            {meals!.map(meal => (
              <li key={meal.idMeal} className={styles.listItem}>
                <Thumbnail
                  to={`/meal/${meal.idMeal}`}
                  label={meal.strMeal}
                  imgSrc={meal.strMealThumb}
                  imgAlt={meal.strMeal}
                  imgWidth={180}
                  imgHeight={180}
                />
              </li>
            ))}
          </ul>
        </>
      );

    case LoadingState.ERROR:
      throw loadError;

    case LoadingState.PENDING:
      return (
        <Center>
          <Spinner />
        </Center>
      );
  }
};

export default Category;
