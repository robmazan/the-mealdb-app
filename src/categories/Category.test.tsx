import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { MealExcerpt } from "../api/theMealDb";
import { LoadingState } from "../shared/useFetch";
import Category from "./Category";

jest.mock("./Thumbnail");

let mockCategoryName: string | undefined = "categoryName";
jest.mock("react-router", () => ({
  useParams() {
    return { categoryName: mockCategoryName };
  }
}));

let mockMeals: MealExcerpt[];
let mockLoadingState: LoadingState;
let mockError: Error | null;

jest.mock("../api/theMealDb", () => ({
  useMealExcerptsOfCategory(categoryName: string) {
    return [mockMeals, mockLoadingState, mockError];
  }
}));

let container: HTMLElement | null = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  if (container) {
    unmountComponentAtNode(container);
  }
  container!.remove();
  container = null;
});

it("shows loading message", async () => {
  mockLoadingState = LoadingState.PENDING;

  await act(async () => {
    render(<Category />, container);
  });

  expect(container!.innerHTML).toContain("Loading");
});

it("renders meals", async () => {
  mockMeals = require("./__mocks__/meals.json");
  mockLoadingState = LoadingState.DONE;

  await act(async () => {
    render(<Category />, container);
  });

  const mealListItems = container!.querySelectorAll("li");
  expect(mealListItems.length).toEqual(mockMeals.length);
});

it("throws error on fetch error", () => {
  mockError = new Error("Fetch error");
  mockLoadingState = LoadingState.ERROR;
  const spy = jest.spyOn(console, "error").mockImplementationOnce(() => {});

  expect(() => render(<Category />, container)).toThrowError(mockError);

  spy.mockRestore();
});
