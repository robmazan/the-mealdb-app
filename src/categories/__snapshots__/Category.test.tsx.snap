// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders meals 1`] = `
<Fragment>
  <h1
    className="pageTitle"
  >
    categoryName
  </h1>
  <ul
    className="list"
  >
    <li
      className="listItem"
      key="1"
    >
      <Thumbnail
        imgAlt="Meal 1"
        imgHeight={180}
        imgSrc="https://example.com/meal1.jpg"
        imgWidth={180}
        label="Meal 1"
        to="/meal/1"
      />
    </li>
    <li
      className="listItem"
      key="2"
    >
      <Thumbnail
        imgAlt="Meal 2"
        imgHeight={180}
        imgSrc="https://example.com/meal2.jpg"
        imgWidth={180}
        label="Meal 2"
        to="/meal/2"
      />
    </li>
    <li
      className="listItem"
      key="3"
    >
      <Thumbnail
        imgAlt="Meal 3"
        imgHeight={180}
        imgSrc="https://example.com/meal3.jpg"
        imgWidth={180}
        label="Meal 3"
        to="/meal/3"
      />
    </li>
  </ul>
</Fragment>
`;
